File: mesa/demo/examples/l2-aggr.c
// Copyright (c) 2004-2020 Microchip Technology Inc. and its subsidiaries.
// SPDX-License-Identifier: MIT


#include <unistd.h>
#include <stdio.h>
#include "cli.h"
#include "example.h"
#include "mscc/ethernet/switch/api.h"

static const char *aggr_help_txt = "\
This is a Link aggregation example configuration.\n\
\n\
Two ports are aggregated. Any other port can be used as ingress port.\n\
The aggregation mode is set to selected through parameter.\n\
";

static struct {
    mesa_aggr_mode_t   mode;
    uint32_t           aggr_no;
} state;

static int aggr_init(int argc, const char *argv[])
{
    mesa_port_no_t     port1 = ARGV_INT("port1", "Is the egress port 1 in the aggregation.");
    mesa_port_no_t     port2 = ARGV_INT("port2", "Is the egress port 2 in the aggregation.");
    uint32_t           aggr_mode = ARGV_OPT_INT("aggr-mode", "Is the selected aggregation mode. Default is 0\n   0 is SMAC\n 1 is DMAC\n   2 is SIP/DIP\n   3 is SPORT/DPORT", 0);
    uint32_t           aggr_no = ARGV_OPT_INT("aggr-no", "Is the selected aggregation number. Default is 1", 1);
    mesa_port_list_t   port_list;
    mesa_aggr_mode_t   mode;

    EXAMPLE_BARRIER(argc);

    memset(&state, 0, sizeof(state));
    state.aggr_no = aggr_no;

    // snippet_begin ex-aggr-enable
    mesa_port_list_clear(&port_list);
    mesa_port_list_set(&port_list, port1-1, TRUE);
    mesa_port_list_set(&port_list, port2-1, TRUE);
    RC(mesa_aggr_port_members_set(NULL, aggr_no, &port_list));

    // snippet_endbegin ex-aggr-mode-set
    RC(mesa_aggr_mode_get(NULL, &mode));
    state.mode = mode;
    switch (aggr_mode) {
    case 0: mode.smac_enable = TRUE; break;
    case 1: mode.dmac_enable = TRUE; break;
    case 2: mode.sip_dip_enable = TRUE; break;
    case 3: mode.sport_dport_enable = TRUE; break;
    }
    RC(mesa_aggr_mode_set(NULL, &mode));
    // snippet_end

    return 0;
}

static int aggr_clean()
{
    mesa_port_list_t   port_list;

    // Disable aggregation
    // snippet_begin ex-aggr-disable
    mesa_port_list_clear(&port_list);
    RC(mesa_aggr_port_members_set(NULL, state.aggr_no, &port_list));
    // snippet_end

    RC(mesa_aggr_mode_set(NULL, &state.mode));

    return 0;
}

static const char* aggr_help()
{
    return aggr_help_txt;
}

EXAMPLE(aggr, aggr_init, NULL, aggr_clean, aggr_help);

File: mesa/demo/test/l2_aggr.rb
#!/usr/bin/env ruby

# Copyright (c) 2004-2020 Microchip Technology Inc. and its subsidiaries.
# SPDX-License-Identifier: MIT

require_relative 'libeasy/et'

$ts = get_test_setup("mesa_pc_b2b_4x")

#---------- Frame testing -----------------------------------------------------

$aggr_no = 1
test "aggr-mode" do
    aggr_fld_table = 
        [ { fld: "smac_enable", smac: ["00:00:00:00:00:04","00:00:00:00:00:05"] },
          { fld: "dmac_enable", dmac: ["00:00:00:00:00:06","00:00:00:00:00:07"] },
          { fld: "sip_dip_enable", sip: ["3.3.3.2", "3.3.3.3"] },
          { fld: "sip_dip_enable", dip: ["4.4.4.4", "4.4.4.5"] },
          { fld: "sport_dport_enable", sport: [16, 17] },
          { fld: "sport_dport_enable", dport: [18, 19] } ]

    # Port index 2 and 3 are aggregated
    $ts.dut.call("mesa_aggr_port_members_set", $aggr_no, port_idx_list_str([2, 3]))

    aggr_fld_table.each do |a|
        [false, true].each do |enabled|
            fld = a[:fld]
            test "aggr-mode, #{fld} = #{enabled}" do
                conf = $ts.dut.call("mesa_aggr_mode_get")
                conf[fld] = enabled
                $ts.dut.call("mesa_aggr_mode_set", conf)
                $ts.dut.p.each_index do |idx_tx|
                    # Send two frames with different aggregation keys
                    base = ethtool_stat($ts)
                    for j in 0..1
                        cmd = "sudo ef tx #{$ts.pc.p[idx_tx]} eth"
                        if (a.key?:dmac)
                            cmd += " dmac #{a[:dmac][j]}"
                        end
                        if (a.key?:smac)
                            cmd += " smac #{a[:smac][j]}"
                        end
                        cmd += " ipv4"
                        if (a.key?:sip)
                            cmd += " sip #{a[:sip][j]}"
                        end
                        if (a.key?:dip)
                            cmd += " dip #{a[:dip][j]}"
                        end
                        cmd += " udp"
                        if (a.key?:sport)
                            cmd += " sport #{a[:sport][j]}"
                        end
                        if (a.key?:dport)
                            cmd += " dport #{a[:dport][j]}"
                        end
                        $ts.pc.run(cmd)
                    end
                    base, diff = ethtool_stat($ts, base)

                    # Check switch Tx (host Rx) counters for each interface
                    prev = 0
                    exp = {}
                    $ts.dut.port_list.each_with_index do |port, idx_rx|
                        name = "#{$ts.pc.p[idx_rx]}"
                        val = diff[name]["rx"]["packets"]
                        cnt = (idx_rx == idx_tx ? 0 :             # Not forwarding back
                               idx_rx < 2 ? 2 :                   # Egress port not aggregated
                               idx_tx > 1 ? 0 :                   # Egress and ingress port aggregated
                               enabled ? 1 :                      # Aggregation field enabled, one frame for each port
                               idx_rx == 2 ? (val == 0 ? 0 : 2) : # 0 frames on one port, 2 frames on the other port
                               prev == 0 ? 2 : 0)
                        exp["#{name}-rx"] = cnt
                        exp["#{name}-tx"] = (idx_rx == idx_tx ? 2 : 0)
                        prev = val
                    end
                    eval_stats(diff, $ts.pc.p, exp)
                end
            end
        end
    end
end

test "aggr-port" do
    aggr_table =
        [ [0,3],
          [1,2,3],
          [0,1,2,3],
          [] ]
    aggr_table.each do |a|
        str = port_idx_list_str(a)
        test "aggr-ports: #{str}" do
            $ts.dut.call("mesa_aggr_port_members_set", $aggr_no, str)
            $ts.dut.port_list.each_index do |idx_tx|
                base = ethtool_stat($ts)
                $ts.pc.run("sudo ef tx #{$ts.pc.p[idx_tx]} eth")
                base, diff = ethtool_stat($ts, base)
                cnt = 0
                $ts.dut.port_list.each_with_index do |port, idx_rx|
                    name = "#{$ts.pc.p[idx_rx]}"
                    val = diff[name]["rx"]["packets"]
                    if (a.include?idx_rx)
                        # Update aggregation counter
                        cnt += val
                    else
                        # Check non-aggregated port counter
                        exp = (idx_tx == idx_rx ? 0 : 1)
                        msg = "port #{port}, Tx expected: #{exp}, value: #{val}"
                        if (exp == val)
                            t_i(msg)
                        else
                            t_e(msg)
                        end
                    end
                end
                # Check aggregation counter
                exp = ((a.include?idx_tx or a.empty?) ? 0 : 1)
                msg = "aggr #{$aggr_no}, Tx expected: #{exp}, value: #{cnt}"
                if (exp == cnt)
                    t_i(msg)
                else
                    t_e(msg)
                end
            end
        end
    end
end
